### PR TITLE
freshrss: init at 1.20.0, tests and module

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Description of changes

All credits go to @etu for starting this PR in #133035! Sadly I was to incompetent to append my changes to your PR. In the end I got frustrated and simply copied the test over, ...

I tested it using the SQLite. So the other Databases are implemented blindly.

The difference between #133035 and this is the separating of the data-folder using the env-var `FRESHRSS_DATA_PATH`.

The module tests are not yet adjusted, but might work since @etu already did some testing.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Stunkymonkey** [comment]
ping @aanderse who did #77354

File: nixos/modules/services/web-apps/freshrss.nix [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      type = types.str;
+      description = "Password for the defaultUser for FreshRSS.";
+      example = "/run/secrets/freshrss";
+    };
+
+    baseUrl = mkOption {
+      type = types.str;
+      description = "Default URL for FreshRSS.";
+      example = "https://freshrss.example.com";
+    };
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:39]: These two options are missing `default`.

File: nixos/modules/services/web-apps/freshrss.nix [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+      };
+
+      host = mkOption {
+        type = types.nullOr types.str;
+        default = "localhost";
+        description = "Database host for FreshRSS.";
+      };
+
+      port = mkOption {
+        type = with types; nullOr (either int str);
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:63]:
```suggestion
        type = with types; nullOr port;
```

File: nixos/modules/services/web-apps/freshrss.nix [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+      description = "Default language for FreshRSS.";
+      example = "de";
+    };
+
+    database = {
+      type = mkOption {
+        type = types.enum [ "sqlite" "pgsql" "mysql" ];
+        default = "sqlite";
+        description = "Database type.";
+        example = "de";
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:53]:
```suggestion
        example = "pgsql";
```

File: nixos/modules/services/web-apps/freshrss.nix [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    in {
+      description = "Set up the state directory for FreshRSS before use";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        User = "freshrss";
+        Group = "freshrss";
+        StateDirectory = "freshrss";
+        WorkingDirectory = cfg.package;
+      };
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:198]: Could you also add more hardening options to this and to the other service as well.

File: nixos/modules/services/web-apps/freshrss.nix [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+    systemd.services.freshrss-updater = {
+      description = "FreshRSS feed updater";
+      after = [ "freshrss-config.service" ];
+      wantedBy = [ "multi-user.target" ];
+      startAt = "*:0/5";
+      environment = {
+        FRESHRSS_DATA_PATH = cfg.dataDir;
+      };
+
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:236]:
```suggestion
```

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -22727,6 +22727,8 @@ with pkgs;
 
   freeradius = callPackage ../servers/freeradius { };
 
+  freshrss = callPackage ../servers/freshrss { };
```
Comment (azahi) [inline pkgs/top-level/all-packages.nix:22730]: Shouldn't it be here instead?
```suggestion
  freshrss = callPackage ../servers/web-apps/freshrss { };
```

File: pkgs/servers/freshrss/default.nix [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      define('DATA_PATH', getenv('FRESHRSS_DATA_PATH'));
+  '';
+
+  installPhase = ''
+    mkdir -p $out
+    cp -vr * $out/
+
+    cp $overrideConfig $out/constants.local.php
+  '';
+
```
Comment (azahi) [inline pkgs/servers/freshrss/default.nix:31]:
```suggestion
```

File: pkgs/servers/freshrss/default.nix [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, lib, fetchFromGitHub, nixosTests, pkgs }:
+
+stdenv.mkDerivation rec {
```
Comment (azahi) [inline pkgs/servers/freshrss/default.nix:3]:
```suggestion
stdenvNoCC.mkDerivation rec {
```

File: pkgs/servers/freshrss/default.nix [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  pname = "FreshRSS";
+  version = "1.19.2";
+
+  src = fetchFromGitHub {
+    owner = "FreshRSS";
+    repo = "FreshRSS";
+    rev = version;
+    sha256 = "sha256-UUDMYgYPvIuL5Jw2fFhJGoJ4KLNUBpk2pPc2UNa+ZSE=";
```
Comment (azahi) [inline pkgs/servers/freshrss/default.nix:11]:
```suggestion
    hash = "sha256-UUDMYgYPvIuL5Jw2fFhJGoJ4KLNUBpk2pPc2UNa+ZSE=";
```

File: pkgs/servers/freshrss/default.nix [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  #passthru.tests = nixosTests.freshrss;
```
Comment (azahi) [inline pkgs/servers/freshrss/default.nix:14]: Is this commented out for a reason?

File: nixos/modules/services/web-apps/freshrss.nix [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        type = types.nullOr types.str;
+        default = null;
+        description = "Database table prefix for FreshRSS.";
+        example = "freshrss";
+      };
+    };
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/freshrss/";
```
Comment (azahi) [inline nixos/modules/services/web-apps/freshrss.nix:98]:
```suggestion
      default = "/var/lib/freshrss";
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/freshrss.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/web-apps/freshrss.nix:39]: The intention is the sysadmin is required to fill in these values. There should be no default.

File: pkgs/servers/freshrss/default.nix (hunk 10 above)
Comment (Stunkymonkey) [reply pkgs/servers/freshrss/default.nix:14]: was commented out while developing.

**Comment by Stunkymonkey** [comment]
any idea, why the manual is not building?

File: nixos/modules/services/web-apps/freshrss.nix (hunk 4 above)
Comment (azahi) [reply nixos/modules/services/web-apps/freshrss.nix:198]: Something like [this](https://github.com/NixOS/nixpkgs/blob/c23131bc34da36f938c6f64de685afbd0f149039/nixos/modules/services/security/endlessh-go.nix#L98-L129). Just run your service through `systemd-analyze security` and patch out some holes.

Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/freshrss.nix]: is this fine 24a1171cf5c7762176f599aa9b87b87169906a11

Comment (azahi) [reply nixos/modules/services/web-apps/freshrss.nix]: Yep, looks much better now.

File: nixos/tests/freshrss.nix [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import ./make-test-python.nix ({ lib, pkgs, ... }: {
+  name = "freshrss";
+  meta.maintainers = with lib.maintainers; [ etu ];
```
Comment (etu) [inline nixos/tests/freshrss.nix:3]: I'm very happy and honored to see this work going on. I'm looking forward to using the module from upstream and migrating my own usage to this!

But I think you @Stunkymonkey should at least co-maintain the tests with me :wink:

**Comment by Stunkymonkey** [comment]
any additional things that need to be done before merging? Can somebody please approve this PR?

**Comment by Lassulus** [comment]
`error: The option `nodes.machine.services.freshrss.baseUrl' is used but not defined.` when running the nixos test

**Comment by Stunkymonkey** [comment]
@Lassulus please review again.

**Comment by pakaca** [comment]
The greader API is not working for me. Mobile apps and newsboat fail to connect to it. 

In the myserver.xxx/api page it shows the API address as myserver.xxx/api/api/greader.php (with /api/ repeated), and reports the configuration test as FAIL. Going manually to myserver/xxx/api/greader.php ghough reports "OK". 

It might very well be just my ignorant self setting it up wrongly

**Comment by Stunkymonkey** [comment]
@pakaca I can confirm, that it shows:
```
Google Reader compatible API

Your API address:
    https://news.buehler.rocks/api/greader.php
Google Reader API configuration test:
    ❌ OK 
```
so i guess this does not work correctly.
could you please open an issue for this.

**Comment by pakaca** [comment]
Ah, ok. You don't get the doublu `api/api/greader.php` though, so I must have some additional config error.

I'll open the issue.

**Pull request merged**